src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/TransPrimitiveClass.java

```diff
@@ -389,14 +389,20 @@ private MethodSymbol getPrimitiveObjectFactory(MethodSymbol init) {
         if (factory != null)
             return factory;
 
-        MethodType factoryType = new MethodType(init.erasure(types).getParameterTypes(),
+        MethodType factoryType = new MethodType(init.type.getParameterTypes(),
                                                 init.owner.type.asValueType(),
                                                 init.type.getThrownTypes(),
                                                 init.owner.type.tsym);
         factory = new MethodSymbol(init.flags_field | STATIC,
                                         names.init,
                                         factoryType,
                                         init.owner);
+        factory.params = init.params;
+        // Re-patch the return type on the erased method type, or code generation will fail
+        factory.erasure_field = new MethodType(init.erasure(types).getParameterTypes(),
+                init.owner.type.asValueType(),
+                init.type.getThrownTypes(),
+                init.owner.type.tsym);
         factory.setAttributes(init);
```

Comment:
Again, I need some help with this change. I understand the part of parameters being carried over, but the dance around recomputation of erasure field - what exactly is the material change there ?? I can see the RecordReading test failing if I backout the changes here (except for factory.params = init.params) but need some explanation - TIA

Reply:
Related to comment above: Since we're now preserving the full type information for `factoryType` and thus `factory`, instead of propagating the erased info, information from the other (preceding) lowerings would get lost. To avoid that, the erasure is explicitly reconstructed, but with a fixed return type (instead of void).

There is a different test which blows up if this isn't reconstructed, it might be `InnerValueNew.java` (I'll check later).